lib.php

```diff
@@ -494,54 +494,78 @@ function block_coursefeedback_get_feedbackname($feedbackid = null) {
 
     return htmlentities($name);
 }
-
 /**
+ * This function gets the amount of votes for each answeroption (1-6) and each question.
+ * It calculates the amount of counted choices, the amount of abstentions and the average rating for each question
+ *
  * @param int $courseid
  * @param string $sort
- * @return array - 2-dimensional array of answers, ordered by question id
+ * @return array - 2-dimensional array of answers, ordered by question id as follows:
+ * [ "<questionid>" =>
+ *      [ "1" => <answercount>, ..., "6" => <answercount>, "average" =>..., "choicessum" =>..., "abstentions" => ... ]]
+ * @throws moodle_exception
  */
-function block_coursefeedback_get_answers($course, $feedbackid, $sort = "questionid") {
-    global $DB, $CFG, $USER;
+function block_coursefeedback_get_qanswercounts($course, $feedbackid) {
+    global $DB;
     $config = get_config("block_coursefeedback");
     $answers = array();
     $course = clean_param($course, PARAM_INT);
 
     if ($course <= 0) {
         throw new moodle_exception("invalidcourseid");
     }
-
-    $questions = block_coursefeedback_get_questions($feedbackid, $config->default_language);
+    // Get all the questions of the feedback
+    $questions = block_coursefeedback_get_questions_by_language($feedbackid, [current_language()]);
     $params = array("fid" => $feedbackid, "course" => $course);
     if (!empty($questions)) {
-        $count = count($questions);
-        foreach (array_keys($questions) as $question) {
-            $params["qid"] = $question;
+        // For each question and each answerpossibility, count the amount of the given answers
+        foreach ($questions as $question) {
+            $choicessum = 0;
+            $avsum = 0;
+            $questionid = $question->questionid;
+            $params["qid"] = $questionid;
             $sql = "SELECT answer,COUNT(*) AS count
-                      FROM {block_coursefeedback_answers}
-                     WHERE coursefeedbackid = :fid 
-                           AND questionid = :qid 
-                           AND course = :course
-                  GROUP BY answer";
-
+                    FROM {block_coursefeedback_answers}
+                    WHERE coursefeedbackid = :fid 
+                        AND questionid = :qid 
+                        AND course = :course
+                    GROUP BY answer";
+
+            // Create array for the question and fill in zeros for each answeroption
+            $answers[$questionid] = array_fill(1, 6, 0);
+
+            $answers[$questionid]['average'] = "x";
+            $answers[$questionid]['choicessum'] = 0;
+            $answers[$questionid]['abstentions'] = 0;
+            // If answers for question exist, replace the zero at the right index and calculate average and choicessum
             if ($results = $DB->get_records_sql($sql, $params)) {
-                $answers[$question] = array();
                 foreach ($results as $answer) {
```

Comment:
Braucht man denn hier das `if` überhaupt noch? Wenn `$results` leer ist, macht die Schleife halt keine Iterationen. Und das, was nach der Schleife kommt, sollte meines Erachtens sowieso immer ausgeführt werden, oder nicht? Dann kannst du dir oben nämlich auch die Initialisierung der inneren Array Werte für `average` und `choicessum` sparen. (Die anderen brauchst du nach wie vor.)

Reply:
Resolve in 8afd5effa1852080a21f283356d08d6aeeaf8d34

Reply:
Resolved in 38d8e093359ee6538a55817e1fa9d81af33f32c3